### ISSUE TITLE
MultiheadAttention output changes if input order is not exactly same

### ISSUE THREAD

**Issue by teucer** [pytorch/pytorch#23525]
## 🐛 Bug

I have trained a transformer model and the results seem to be wrong

## To Reproduce

1. Train a model using the multi head attention (encoder only) 
2. construct vector x = [x1,x2] and xx = [x3,x1] where x1, x2, and x3 are two inputs 
3. calculate y = transformer(x) and yy = transformer(xx)

## Expected behavior

y[0] = yy[1]

Believe the issue due to the reshaping of q,k,v in torch\nn\modules\activation.py (lines 804-808) and as a consequence the softmax at line 860 in the same file. 

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): windows
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): na
 - Python version: 3.7.3
 - CUDA/cuDNN version: na
 - GPU models and configuration: na
 - Any other relevant information:

## Additional context

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by andreaskoepf** [comment]
Are you talking about self-attention?
Could you please post a minimized example to reproduce the problem?

If two batches have sequences with the same embeddings (in different order) the output shoud be identical for self-atttention (dot-product scores before softmax are the same) only in a different order, e.g. the following test should hold:

```
x = torch.rand(2, 2, 16)  # len, batch_size, d_model  
x[0,1] = x[1,0]                # copy elements from first batch but swap order
x[1,1] = x[0,0]
attn = nn.MultiheadAttention(16, 4)
y,_ = attn(x,x,x)
assert(torch.equal(y[0,0], y[1,1]) and torch.equal(y[0,1], y[1,0]))
```

But if you introduce a completely new element `x3` the dot-product score sum will differ and therefore also the softmax-result, therefore I do not think that `y[0] = yy[1]` is the expected behavior.

**Comment by vincentqb** [comment]
Thank you for reporting this. As @andreaskoepf mentioned, can you provide a minimal code that reproduces this issue?

**Comment by taras-janea** [comment]
Closed the issue due to the retention period and not being able to reproduce it.
